gmail-to-salesforce-lead/main.bal

```diff
@@ -0,0 +1,193 @@
+                leads.push(lead);
+            }
+        };
+
+        check addLeadsToSalesforce(leads);
+
+        runtime:sleep(600);
+    }
+}
+
+function getEmails(string label) returns Email[]|error {
+    gmail:Client gmailClient = check new ({auth: {token: gmailAccessToken}});
+
+    string[] labelIdsToMatch = check getLabelIds(gmailClient, [label]);
+    if (labelIdsToMatch.length() == 0) {
+        error e = error("Unable to find any labels to match.");
+        return e;
+    }
+
+    gmail:MailThread[] matchingMailThreads = check getMatchingMailThreads(gmailClient, labelIdsToMatch);
+    removeLabels(gmailClient, matchingMailThreads, labelIdsToMatch);
+
+    gmail:Message[] matchingEmails = getMatchingEmails(gmailClient, matchingMailThreads);
+
+    Email[] emails = [];
+    from gmail:Message message in matchingEmails
+    do {
+        Email|error email = parseEmail(message);
+        if email is Email {
+            emails.push(email);
+        }    
+    };
+    
+    return emails;
+}
+
+function getLabelIds(gmail:Client gmailClient, string[] labelsToMatch) returns string[]|error {
+    gmail:LabelList labelList = check gmailClient->listLabels("me");
+
+    return from gmail:Label label in labelList.labels
+        where labelsToMatch.indexOf(label.name) != ()
+        select label.id;
+}
+
+function getMatchingMailThreads(gmail:Client gmailClient, string[] labelIdsToMatch) returns gmail:MailThread[]|error {
+    gmail:MsgSearchFilter searchFilter = {
+        includeSpamTrash: false,
+        labelIds: labelIdsToMatch
+    };
+
+    stream<gmail:MailThread, error?>|error mailThreadStream = gmailClient->listThreads(filter = searchFilter);
+    if mailThreadStream is error {
+        return mailThreadStream;
+    }
+
+    return check from gmail:MailThread mailThread in mailThreadStream
+        select mailThread;
+}
+
+function removeLabels(gmail:Client gmailClient, gmail:MailThread[] mailThreads, string[] labelIds) {
+    from gmail:MailThread mailThread in mailThreads
+    do {
+        gmail:MailThread|error removeLabelResponse = gmailClient->modifyThread(mailThread.id, [], labelIds);
+        if removeLabelResponse is error {
+            log:printError("An error occured in removing the labels from the thread.", removeLabelResponse, removeLabelResponse.stackTrace(), threadId = mailThread.id, labelIds = labelIds);
+        }
+    };
+}
+
+function getMatchingEmails(gmail:Client gmailClient, gmail:MailThread[] mailThreads) returns gmail:Message[] {
+    gmail:Message[] messages = [];
+    _ = from gmail:MailThread mailThread in mailThreads
+        do {
+            gmail:MailThread|error response = gmailClient->readThread(mailThread.id);
+            if response is error {
+                log:printError("An error occured while reading the email.", response, response.stackTrace(), threadId = mailThread.id);
+            } else {
+                messages.push((<gmail:Message[]>response.messages)[0]);
```

Comment:
Casts panic if the value don't belong to the target type. ensureType or an explicit is check will be better.